src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java

```diff
@@ -25,48 +25,95 @@
 package jenkins.metrics.impl;
 
 import hudson.ExtensionList;
+import hudson.model.FreeStyleBuild;
 import hudson.model.FreeStyleProject;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.TimeUnit;
+
+import hudson.model.Run;
+import hudson.model.queue.QueueTaskFuture;
 import jenkins.metrics.api.Metrics;
 import jenkins.metrics.api.QueueItemMetricsEvent;
 import jenkins.metrics.api.QueueItemMetricsListener;
+import org.junit.ClassRule;
+import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
+import org.jvnet.hudson.test.BuildWatcher;
 import org.jvnet.hudson.test.JenkinsRule;
 import org.jvnet.hudson.test.SleepBuilder;
 import org.jvnet.hudson.test.TestExtension;
 
+import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.emptyCollectionOf;
 import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThan;
 import static org.hamcrest.Matchers.notNullValue;
-import static org.junit.Assert.assertThat;
 
 public class JenkinsMetricProviderImplTest {
     @Rule
     public JenkinsRule j = new JenkinsRule();
 
+    @ClassRule
+    public static BuildWatcher w = new BuildWatcher();
+
+    @Test
+    public void given__a_job_which_is_cancelled() throws Exception {
+        MyListener listener = ExtensionList.lookup(QueueItemMetricsListener.class).get(MyListener.class);
+        assertThat(listener, notNullValue());
+
+        FreeStyleProject job = j.createProject(FreeStyleProject.class);
+
+        QueueTaskFuture<FreeStyleBuild> taskFuture = job.scheduleBuild2(10);
+        taskFuture.cancel(true);
+
+        try{
+            j.waitForCompletion(taskFuture.waitForStart());
+        }catch (Throwable t){
+            //CancellationException expected
+        }
+
+        //wait for completion
+        while (!listener.state.toString().contains("C")){
+            Thread.sleep(10);
+        }
+
+        assertThat(listener.state.toString(), containsString("Q"));
```

Comment:
Could you please assert with equal? It leaves the test more comprehensive and avoid failures in case we add some new state on the future with this letter. The same for the other cases.